content/posts/wasm-as-a-platform-for-abstraction.md

```diff
@@ -0,0 +1,1390 @@
+    message: WasmPtr<u8, Array>,
+    message_len: i32,
+) -> i32 {
+    // Note: We can't directly accept the Level enum here because out-of-range
+    // enum variants are UB
+    let level = match level {
+        LOG_ERROR => Level::Error,
+        LOG_WARN => Level::Warn,
+        LOG_INFO => Level::Info,
+        LOG_DEBUG => Level::Debug,
+        LOG_TRACE => Level::Trace,
+        _ => Level::Debug,
+    };
+    let filename = file.get_utf8_string(ctx.memory(0), file_len as u32);
+    let message = message
+        .get_utf8_string(ctx.memory(0), message_len as u32)
+        .unwrap_or_default();
+gt
+    unsafe {
+        try_with_env!(
+            ctx,
+            // unfortunately constructing a log and using it needs to be in a
+            // single statement because lifetimes
+            // https://users.rust-lang.org/t/using-format-args-and-log-builder/22695
+            log(&Record::builder()
+                .level(level)
+                .file(filename.as_deref())
+                .line(Some(line as u32))
+                .args(format_args!("{}", message))
+                .build()),
+            "Logging failed"
+        );
+    }
+
+    WASM_SUCCESS
+}
+```
+
+Implementing the other host functions follows the same steps. After writing a
+couple of these function "trampolines", it goes from being a scary `unsafe`
+task to a mechanical job of translating arguments and error values.
+
+{{% expand "A big wall of code that translates arguments and error values." %}}
+```rust
+// src/lib.rs
+
+impl TryFrom<Value> for bool {
+    type Error = Error;
+
+    fn try_from(other: Value) -> Result<bool, Self::Error> {
+        match other {
+            Value::Bool(b) => Ok(b),
+            _ => Err(Error::BadVariableType),
+        }
+    }
+}
+
+impl TryFrom<Value> for i32 {
+    type Error = Error;
+
+    fn try_from(other: Value) -> Result<i32, Self::Error> {
+        match other {
+            Value::Integer(i) => Ok(i),
+            _ => Err(Error::BadVariableType),
+        }
+    }
+}
+
+impl TryFrom<Value> for f64 {
+    type Error = Error;
+
+    fn try_from(other: Value) -> Result<f64, Self::Error> {
+        match other {
+            Value::Float(f) => Ok(f),
+            _ => Err(Error::BadVariableType),
+        }
+    }
+}
+
+impl From<bool> for Value {
+    fn from(b: bool) -> Value { Value::Bool(b) }
+}
+
+impl From<i32> for Value {
+    fn from(i: i32) -> Value { Value::Integer(i) }
+}
+impl From<f64> for Value {
+    fn from(d: f64) -> Value { Value::Float(d) }
+}
+
+fn wasm_read_input(
+    ctx: &mut Ctx,
+    address: u32,
+    buffer: WasmPtr<u8, Array>,
+    buffer_len: i32,
+) -> i32 {
+    let mut temp_buffer = vec![0; buffer_len.try_into().unwrap()];
+
+    unsafe {
+        try_with_env!(
+            ctx,
+            read_input(address.try_into().unwrap(), &mut temp_buffer[..]),
+            "Unable to read the input"
+        );
+    }
+
+    wasm_deref!(with ctx, *buffer = for byte in temp_buffer);
```

Comment:
for byte in temp_buffer. Thats elegant, does it return a slice?

Reply:
The `WasmPtr` type has this extra type "tag" to differentiate between a pointer to a single item, and an array. Kinda like how `char *` may point to a single character or a string of characters.

When you have a `WasmPtr<T, Array>` it gets this [`deref()`](https://docs.rs/wasmer-runtime/0.11.0/wasmer_runtime/struct.WasmPtr.html#method.deref-1) method which will return `Option<&'a [Cell<T>]>`. 

I made the macro match `*$ptr:ident = for $item:ident in $collection:expr` to help remember what `wasm_deref!()` is trying to do without having to read the expanded code (error handling is kinda clunky because we can't use `?`). If you look at the macro definition we deref to get the `Option<&[Cell<T>]>` then if it's `Some` we'll iterate `for i in 0..collection.len()` and do the equivalent of `buffer[i] = collection[i]`.